src/cata_utility.cpp

```diff
@@ -18,9 +18,25 @@
 #include <string>
 #include <locale>
 
+static double pow10( unsigned int n )
+{
+    double ret = 1;
+    double tmp = 10;
+    while( n ) {
+        if( n & 1 ) {
+            ret *= tmp;
+        }
+        tmp *= tmp;
+        n >>= 1;
+    }
+    return ret;
+}
+
```

Comment:
internal to the function you are dealing with only integer values, so you can use an int or int64 to accumulate the result doing integer arithmetic (faster than float/double precision ops). This works for an int64 while `n <= 18` holds. In any case, you can eliminate the branch at `if( n & 1 )`. With `n <= 9` an int32 can be used as the return type.

```
// This is just for testing different result types
typedef double power_result_t;
power_result_t pow10( unsigned int n ){
    power_result_t result = 1;
    power_result_t base = 10;
    // No harm in going through at least once. If (n) is 0 the result will be 1.
    do{
        // Branch elimination work
        const power_result_t nao = n & 1;
        const power_result_t base_nao = base * nao;
        const power_result_t omnao = 1 - nao;
        // Branchless evaluation of "if( n & 1 ){ ret *= tmp; }
        // base_mul is either base_nao or 1
        const power_result_t base_mul = base_nao + omnao;
        result *= base_mul;
        n >>= 1;
        base *= base;
    }while (n);

    return result;
}
```
Possibly harder to read, and may not be worth using since this gets called infrequently. Modifications are entirely geared towards eliminating the cost of a missed branch prediction in `if( n & 1 )`.

Reply:
As you said this gets called infrequently, so no need to optimized it for now. Anyway, nice work!